engine/wait_for_process_completion_channel.go

```diff
@@ -0,0 +1,150 @@
+
+package engine
+
+import (
+	"fmt"
+	"github.com/xcherryio/xcherry/common/log"
+	"time"
+)
+
+type WaitForProcessCompletionChannelsPerShardImpl struct {
+	shardId int32
+	logger  log.Logger
+
+	processor ImmediateTaskProcessor
+
+	// processExecutionId : channel
+	channelMap map[string]chan string
+	// processExecutionId : a list of timestamps of when the waiting requests were created
+	waitingRequestCreatedAt map[string][]int64
+}
+
+func NewWaitForProcessCompletionChannelsPerShardImplImpl(
+	shardId int32, logger log.Logger, processor ImmediateTaskProcessor) WaitForProcessCompletionChannelsPerShard {
+	return &WaitForProcessCompletionChannelsPerShardImpl{
+		shardId: shardId,
+		logger:  logger,
+
+		processor: processor,
+
+		channelMap:              map[string]chan string{},
+		waitingRequestCreatedAt: map[string][]int64{},
+	}
+}
+
+func (w *WaitForProcessCompletionChannelsPerShardImpl) Start() {
+	w.processor.AddWaitForProcessCompletionChannels(w.shardId, w)
+}
+
+func (w *WaitForProcessCompletionChannelsPerShardImpl) Stop() {
+	w.processor.RemoveWaitForProcessCompletionChannels(w.shardId)
+
+	var procIds []string
+
+	for procId := range w.channelMap {
+		procIds = append(procIds, procId)
+	}
+
+	for _, procId := range procIds {
+		w.Signal(procId, WaitForProcessCompletionResultStop)
+	}
+}
+
+func (w *WaitForProcessCompletionChannelsPerShardImpl) Add(processExecutionId string) chan string {
+	w.logger.Info(fmt.Sprintf("Add process execution completion waiting request for %s in shard %d",
+		processExecutionId, w.shardId))
+
+	channel, ok := w.channelMap[processExecutionId]
+	if !ok {
+		channel = make(chan string)
+		w.channelMap[processExecutionId] = channel
+	}
+
+	w.waitingRequestCreatedAt[processExecutionId] = append(w.waitingRequestCreatedAt[processExecutionId], w.now())
+
+	w.terminationCheck(processExecutionId)
+
+	return channel
+}
+
+func (w *WaitForProcessCompletionChannelsPerShardImpl) Signal(processExecutionId string, result string) {
+	channel, ok := w.channelMap[processExecutionId]
+	if !ok {
+		return
+	}
+
+	count := len(w.waitingRequestCreatedAt[processExecutionId])
+
+	for i := 0; i < count; i++ {
+		select {
+		case channel <- result:
+			w.logger.Info(fmt.Sprintf("Signal process execution completion waiting result %d for %s: %s",
+				i, processExecutionId, result))
+		default:
+			w.logger.Info(fmt.Sprintf("Not signal process execution completion waiting result %d for %s: %s",
+				i, processExecutionId, result))
+		}
+	}
+
+	w.waitingRequestCreatedAt[processExecutionId] = []int64{}
+
+	go func() {
+		// sleep 3 seconds before close the channel
+		time.Sleep(time.Second * 3)
+
+		w.cleanup(processExecutionId)
+	}()
+}
+
+func (w *WaitForProcessCompletionChannelsPerShardImpl) terminationCheck(processExecutionId string) {
+	// To check if the channel should be closed after a certain time period
+	go func() {
+		time.Sleep(time.Second * time.Duration(DEFAULT_WAIT_FOR_TIMEOUT_MAX+3))
```

Comment:
This is a bit wasteful and hacky to use a background thread + timer. 

A better/clean and formal way is to let/ensure every waiting request call the cleanup at the end.

Reply:
Currently, the terminationCheck is only called after adding a request, we need to wait for the timeout to perform the cleanup. One request will call once of the method only.